File: old/JNibbles/src/main/java/org/lagash/DiegoGBehavior.java
/**
 * Copyright (C) 2017 DiegoG.
 */
package org.lagash;

import org.lagash.nibbles.IPosition;
import org.lagash.nibbles.ISnake;
import org.lagash.nibbles.ISpace;
import org.lagash.nibbles.Direction;
import org.lagash.nibbles.ISnakeBehavior;

import java.util.List;

public class DiegoGBehavior implements ISnakeBehavior {
    public Direction changeDirection(ISnake snake, ISpace space, List<ISnake> snakes) {
        // Si no me choco adelante, sigo igual
        IPosition pos = snake.moveNew();
        if(pos.isValid(space.getTopX(), space.getTopY())
            && (space.get(pos.getX(), pos.getY()) == 0)) {
            return snake.getDirection();                
        }

        //Busco nueva dirección clockwise para no chocarme
        for(int i = 0; i < 4; i++) {
            Direction dir = Direction.values()[i];
            pos = snake.moveNew(dir);
            if(!pos.isValid(space.getTopX(), space.getTopY())) {
                continue;
            }
            if(space.get(pos.getX(), pos.getY()) > 0) {
                continue;
            }
            return dir;
        }
        return snake.getDirection();
    }
}

File: old/DiegoG/DiegoGBehavior.cs
namespace Nibbles
{
    using System.Collections.Generic;

    public class DiegoGBehavior : ISnakeBehavior
    {
        public Direction ChangeDirection(ISnake snake, ISpace space, IReadOnlyCollection<ISnake> snakes)
        {
            // Si no me choco adelante, sigo igual
            IPosition pos = snake.MoveNew();
            if(pos.IsValid(space.TopX, space.TopY)
               && (space[pos.X, pos.Y] == 0))
            {
                return snake.Direction;                
            }
            
            //Busco nueva dirección clockwise para no chocarme
            foreach(Direction dir in new[]{0,1,2,3})
            {
                pos = snake.MoveNew(dir);
                if(!pos.IsValid(space.TopX, space.TopY))
                {
                    continue;
                }
                if(space[pos.X, pos.Y] > 0)
                {
                    continue;
                }
                return dir;
            }
            return snake.Direction;
            
            //TODO: no meterse en callejones
        }
    }
}